### PR TITLE
Extending native integration tests for service runtime [ECR-3011].

### PR DESCRIPTION
**Pull request by skletsun**
## Overview
Integration tests for service runtime are extended with fake service and testing utils for artifacts generation.

---
See: https://jira.bf.local/browse/ECR-3011

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Change is covered by automated [tests](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#tests)
- [x] The [coding guidelines](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [x] Public API has Javadoc
- [x] Method preconditions are checked and documented in the Javadoc of the method
- [x] Changelog is updated if needed (in case of notable or breaking changes)
- [x] The continuous integration build passes

### CODE REVIEW

File: exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs [hunk 1]
```diff
@@ -17,14 +17,100 @@
...
+    runtime.load_artifact(&artifact_path);
+}
+
+#[test]
+#[should_panic(expected = "Unable to load artifact")]
+fn load_artifact_twice() {
+    let runtime = get_runtime();
+    let artifact_path = create_service_artifact_valid(runtime.get_executor());
+
+    runtime.load_artifact(&artifact_path);
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs]: This test is especially susceptible to this issue: https://errorprone.info/bugpattern/TestExceptionChecker

File: exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs [hunk 2]
```diff
@@ -110,32 +140,13 @@ fn create_service_for_unknown_artifact() {
...
+    JavaServiceRuntime::new(VM.clone(), get_port())
 }
 
-// Returns a guard to the service runtime that helps executing tests in turn.
-fn get_runtime() -> MutexGuard<'static, JavaServiceRuntime> {
-    SERVICE_RUNTIME.lock()
+// Provides port number that is unique within the scope of tests module.
+fn get_port() -> i32 {
+    static PORT: AtomicUsize = ATOMIC_USIZE_INIT;
+    PORT.fetch_add(1, Ordering::SeqCst) as i32 + 6300
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs]: Why are these complications? Can't we just pass 0 so that the server picks any available port?

File: exonum-java-binding/core/rust/src/runtime/service_factory_adapter.rs [hunk 3]
```diff
@@ -58,7 +58,12 @@ impl JavaServiceFactoryAdapter {
                 system_lib_path: Some(absolute_library_path()),
             };
 
-            let runtime = JavaServiceRuntime::new(config, internal_config);
+            let vm = JavaServiceRuntime::create_java_vm(
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/rust/src/runtime/service_factory_adapter.rs]: Why this change in the production code? Isn't this method supposed to be test-only?

File: exonum-java-binding/core/rust/src/runtime/java_service_runtime.rs [hunk 4]
```diff
@@ -42,29 +42,27 @@ pub struct JavaServiceRuntime {
...
-    /// Creates new runtime from provided config.
+    /// Creates new runtime for given JVM and port.
     ///
-    /// There can be only one `JavaServiceRuntime` instance at a time.
-    pub fn new(config: Config, internal_config: InternalConfig) -> Self {
-        let java_vm =
-            Self::create_java_vm(&config.jvm_config, &config.runtime_config, internal_config);
-        let executor = MainExecutor::new(Arc::new(java_vm));
-        let service_runtime = Self::create_service_runtime(config.runtime_config, executor.clone());
+    /// There can be only one JVM per process.
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/rust/src/runtime/java_service_runtime.rs]: This comment does not work here.

File: exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs [hunk 5]
```diff
@@ -19,21 +19,51 @@ extern crate integration_tests;
...
+    atomic::{AtomicUsize, Ordering, ATOMIC_USIZE_INIT},
+    Arc,
+};
 
 lazy_static! {
-    // Reference to the JavaServiceRuntime is wrapped with Mutex in order to execute tests (more
-    // precise - Java code for artifacts loading and services creation) sequentially.
-    // We use Mutex from the `parking_lot` crate because it doesn't get poisoned in case of panic.
-    static ref SERVICE_RUNTIME: Mutex<JavaServiceRuntime> = Mutex::new(create_runtime());
+    static ref VM: Arc<JavaVM> = {
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs]: Why are all these empty parameters needed? Is there _anything_ happening except passing the classpath?

**Comment by dmitry-timofeev** [comment]
> LGTM except for the issue with load_artifact_twice

Which issue?

**Comment by vitvakatu** [comment]
You left a comment: 

>This test is especially susceptible to this issue: https://errorprone.info/bugpattern/TestExceptionChecker

File: exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs (hunk 2 above)
Comment (skletsun) [reply exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs]: Done, thanks for the idea, somehow I forgot it's possible to do it that way...

File: exonum-java-binding/core/rust/src/runtime/service_factory_adapter.rs (hunk 3 above)
Comment (skletsun) [reply exonum-java-binding/core/rust/src/runtime/service_factory_adapter.rs]: Rewritten with an additional method for integration tests.

File: exonum-java-binding/core/rust/src/runtime/java_service_runtime.rs (hunk 4 above)
Comment (skletsun) [reply exonum-java-binding/core/rust/src/runtime/java_service_runtime.rs]: Reverted

File: exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs (hunk 5 above)
Comment (skletsun) [reply exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs]: Creation of JVM with configs is moved to a separate test in a separate file.

File: exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs [hunk 6]
```diff
@@ -109,10 +75,16 @@ fn load_nonexistent_artifact() {
...
-    runtime.load_artifact(&artifact_path);
+    let res = catch_unwind(|| {
+        let runtime = get_runtime();
+        let artifact_path = create_service_artifact_valid(runtime.get_executor());
+        runtime.load_artifact(&artifact_path);
+        (runtime, artifact_path)
+    });
+    assert!(res.is_ok());
+
+    let (runtime, artifact_path) = res.unwrap();
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs]: Wouldn't it be more natural and easier to read if the approach earlier recommended was used? Why these unnatural unwraps and redundant checks of the code that is _not_ being tested?

```
fn assert_panics(error_message_pattern: String, f: Callable) {
  let res = catch_unwind(f.call());
  assert(res.is_err());
  assert_matches|equals|starts_with(res.get_message(), error_message_pattern);
}
```

Comment (skletsun) [reply exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs]: Now I get what you mean. Definitely, it looks better.

File: exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs [hunk 7]
```diff
@@ -65,54 +65,80 @@ fn load_two_services() {
...
+    );
 }
 
 // Creates a new instance of JavaServiceRuntime for same JVM.
 fn get_runtime() -> JavaServiceRuntime {
     JavaServiceRuntime::create_with_jvm(VM.clone(), 0)
 }
+
+// Asserts that given closure panics while executed.
+fn assert_panics<F, R>(err_pattern: &str, f: F)
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs]: Is it a pattern or, as the code below suggests, a sub-string?

Comment (skletsun) [reply exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs]: Actually, it's a substring, but it's called "err_pattern" because in fact it gets passed to the underlying functions that expect a `Pattern` entity and treated as a pattern inside a string. Do you want me to rename it to smth like `err_substring`?

Comment (dmitry-timofeev) [reply exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs]: I am not sure I understand: can a user pass a pattern (a glob, a regex?), or only a substring?

Comment (skletsun) [reply exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs]: Ah, no, substring only, no patterns in term of regex.

Comment (dmitry-timofeev) [reply exonum-java-binding/core/rust/integration_tests/tests/service_runtime.rs]: Then please name accordingly

**Comment by dmitry-timofeev** [comment]
@dip56 Would you please test this on JDK 12? The changes from this PR used to fail in native ITs (when Travis _was_ able to install it).

**Comment by dmitry-timofeev** [comment]
Will see if ITs will pass on 12 now, when it is back on Travis
